src/Storages/ObjectStorage/DataLakes/DeltaLake/TableSnapshot.cpp

```diff
@@ -0,0 +1,262 @@
+#include "config.h"
+
+#if USE_DELTA_KERNEL_RS
+
+#include <Storages/ObjectStorage/DataLakes/DeltaLake/TableSnapshot.h>
+#include <Storages/ObjectStorage/DataLakes/DeltaLake/ObjectInfoWithPartitionColumns.h>
+#include <Storages/ObjectStorage/DataLakes/DeltaLake/parseFieldFromString.h>
+#include <Core/Types.h>
+#include <Core/NamesAndTypes.h>
+#include <Core/Field.h>
+#include <Common/Exception.h>
+#include <Common/logger_useful.h>
+#include "getSchemaFromSnapshot.h"
+#include "KernelUtils.h"
+
+namespace fs = std::filesystem;
+
+namespace DB::ErrorCodes
+{
+    extern const int LOGICAL_ERROR;
+}
+
+namespace DeltaLake
+{
+
+class TableSnapshot::Iterator final : public DB::IObjectIterator
+{
+public:
+    Iterator(
+        const KernelExternEngine & engine_,
+        const KernelSnapshot & snapshot_,
+        const std::string & data_prefix_,
+        const DB::NamesAndTypesList & schema_,
+        const DB::Names & partition_columns_,
+        DB::ObjectStoragePtr object_storage_,
+        LoggerPtr log_)
+        : scan(KernelUtils::unwrapResult(ffi::scan(snapshot_.get(), engine_.get(), /* predicate */{}), "scan"))
+        , scan_data_iterator(KernelUtils::unwrapResult(
+            ffi::kernel_scan_data_init(engine_.get(), scan.get()),
+            "kernel_scan_data_init"))
+        , data_prefix(data_prefix_)
+        , schema(schema_)
+        , partition_columns(partition_columns_)
+        , object_storage(object_storage_)
+        , log(log_)
+    {
+    }
+
+    size_t estimatedKeysCount() override
+    {
+        /// For now do the same as StorageObjectStorageSource::GlobIterator.
+        /// TODO: is it possible to do a precise estimation?
```

Comment:
No idea, object storages doesn't provide any way to get this metadata. Maybe only for Data Lakes/Via Catalog...